
# Instructions

this will earn level 2 for **construct**

Your goal is to build and prepare two ready to analyze datasets. You will submit a notebook that describes how you built and prepared each dataset.

* Each finished dataset must be produced from two or more tables.

* At least one must come from an sqlite database, either by merging results from multiple queries or multiple tables.

* You should use at least three different merges and one concatenate

Your completed datasets should have:

* column names that are well formatted (only lowercase letters, numbers and _)

* an added column that is derived from one or more other columns (string operation or calculation)

For each dataset, pose one question that could not be answered from the input data files as provided and demonstrate how to answer it with the dataset you built. This could be something that can be answered with using only shape of the merged data, but if you need summarize and visualize level 2 achievements, you should use more statistics and plots.

Your notebooks must be in the top level of the repository, not in a subfolder.
Additional Achievements:

if you already earned prior achievements you can ignore the following

To earn level 2 for **prepare**, one of your analyses must use datasets with missing values and one must be provided as excel files with merged columns (for example from NCES). You may use one dataset with both merged columns and missing data or one of each. You must also use datasets that have column names that need repair.
For the merged column data, either before or after merging, you must additionally:

* create separate separate tables for original and aggregate values (eg percentages or sums that can be recovered from the other columns)

* unstack all levels of the data to create a single level index over the columns* create a database with the tables


![mobile](mobile.jpg)

# Description of Datasets
These data are etracted from [KillBiller application](http://www.killbiller.com/) and some downloaded data, contained in three CSV files:

user_usage.csv : A first dataset containing users monthly mobile usage statistics
user_device.csv : A second dataset containing details of an individual “use” of the system, with dates and device information.
android_devices.csv : A third dataset with device and manufacturer data, which lists all Android devices and their model code, obtained from [Google](https://support.google.com/googleplay/answer/1727131?hl=en-GB).

The main question which we are seeking to answer in this project will be: How usage of mobile services is affected by the manufacturer of user's mobile?

[Reference](https://www.shanelynn.ie/merge-join-dataframes-python-pandas-index-1/)


**What is a merge or join of two dataframes?**

Merging two datasets is the process of bringing two datasets together into one, and aligning the rows from each based on common attributes or columns.

Below, we can see the Venn diagrams which are commonly used to exemplify the different merge and join types.

![Merge/Join types as used in Pandas, R, SQL, and other data-orientated languages and libraries.](https://shanelynnwebsite-mid9n9g1q9y8tt.netdna-ssl.com/wp-content/uploads/2017/03/join-types-merge-names.jpg)

[Reference](https://www.shanelynn.ie/merge-join-dataframes-python-pandas-index-1/)

## Importing Required Libraries

In [1]:
import pandas as pd

## Loading Datasets

In [2]:
#Loading "devices_df" dataset
devices_df = pd.read_csv("data/mobile/android_devices.csv")

In [3]:
#Loading "user_df" dataset
user_df = pd.read_csv("data/mobile/user_device.csv")

In [4]:
#Loading "usage_df" dataset
usage_df = pd.read_csv("data/mobile/user_usage.csv")

## Shape of Datasets

In [5]:
#shape of "devices_df" dataset
devices_df.shape 

(14546, 4)

In [6]:
#shape of "user_df" dataset
user_df.shape

(272, 6)

In [7]:
#shape of "usage_df" dataset
usage_df.shape

(240, 4)

## Overview on three datasets

In [8]:
devices_df.head()

Retail Branding Marketing Name    Device                      Model
0             NaN            NaN    AD681H  Smartfren Andromax AD681H
1             NaN            NaN     FJL21                      FJL21
2             NaN            NaN       T31              Panasonic T31
3             NaN            NaN  hws7721g         MediaPad 7 Youth 2
4              3Q        OC1020A   OC1020A                    OC1020A

In [9]:
user_df.head()

use_id  user_id platform  platform_version     device  use_type_id
0   22782    26980      ios              10.2  iPhone7,2            2
1   22783    29628  android               6.0    Nexus 5            3
2   22784    28473  android               5.1   SM-G903F            1
3   22785    15200      ios              10.2  iPhone7,2            3
4   22786    28239  android               6.0  ONE E1003            1

In [10]:
usage_df.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id
0                    21.97                    4.82     1557.33   22787
1                  1710.08                  136.88     7267.55   22788
2                  1710.08                  136.88     7267.55   22789
3                    94.46                   35.17      519.12   22790
4                    71.59                   79.26     1557.33   22792

## Left Merge

When we render left merge on two datasets, we are keeping all of the rows and values from the left dataframe, ("user_usage" in the following example). Rows of right dataframe will not be dropped if there is a match in the merge variable in the right dataframe, and NaN values will be in the result where not.

we put the "usage_df" as the left dataset, and "user_df" as the right one.

In [11]:
#Merging on the left dataset
merged_left1 = pd.merge(usage_df, user_df, on = 'use_id', how = 'left' )

In [12]:
#overview on the "merged_left1" dataset
merged_left1.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   
3                    94.46                   35.17      519.12   22790   
4                    71.59                   79.26     1557.33   22792   

   user_id platform  platform_version    device  use_type_id  
0  12921.0  android               4.3  GT-I9505          1.0  
1  28714.0  android               6.0  SM-G930F          1.0  
2  28714.0  android               6.0  SM-G930F          1.0  
3  29592.0  android               5.1     D2303          1.0  
4  28217.0  android               5.1  SM-G361F          1.0

In [13]:
#shape of "merged_left1" dataset
merged_left1.shape

(240, 9)

## Right Merge

In [14]:
#merging on the right dataset
merged_right1 = pd.merge(usage_df, user_df, on = 'use_id', how = 'right' )

In [15]:
#overviw on the merged_right1 dataset
merged_right1.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                      NaN                     NaN         NaN   22782   
1                      NaN                     NaN         NaN   22783   
2                      NaN                     NaN         NaN   22784   
3                      NaN                     NaN         NaN   22785   
4                      NaN                     NaN         NaN   22786   

   user_id platform  platform_version     device  use_type_id  
0    26980      ios              10.2  iPhone7,2            2  
1    29628  android               6.0    Nexus 5            3  
2    28473  android               5.1   SM-G903F            1  
3    15200      ios              10.2  iPhone7,2            3  
4    28239  android               6.0  ONE E1003            1

In [16]:
#shape of "merged_right1" dataset
merged_right1.shape

(272, 9)

## Outer Merge
outer merge uses union of keys from both frames.

In [17]:
#outer-merging
merged_outer1 = pd.merge(usage_df, user_df, on = 'use_id', how = 'outer' )

In [18]:
#overview on merged_outer1
merged_outer1.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   
3                    94.46                   35.17      519.12   22790   
4                    71.59                   79.26     1557.33   22792   

   user_id platform  platform_version    device  use_type_id  
0  12921.0  android               4.3  GT-I9505          1.0  
1  28714.0  android               6.0  SM-G930F          1.0  
2  28714.0  android               6.0  SM-G930F          1.0  
3  29592.0  android               5.1     D2303          1.0  
4  28217.0  android               5.1  SM-G361F          1.0

In [19]:
#shape of merged_outer1
merged_outer1.shape

(353, 9)

## Inner Merge
Inner Merge uses intersection of keys from both frames

In [20]:
#inner-merging(inner is default mode)
merged_inner1 = pd.merge(usage_df, user_df, on = 'use_id' )

In [21]:
#overview on merged_inner1
merged_inner1.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   
3                    94.46                   35.17      519.12   22790   
4                    71.59                   79.26     1557.33   22792   

   user_id platform  platform_version    device  use_type_id  
0    12921  android               4.3  GT-I9505            1  
1    28714  android               6.0  SM-G930F            1  
2    28714  android               6.0  SM-G930F            1  
3    29592  android               5.1     D2303            1  
4    28217  android               5.1  SM-G361F            1

In [22]:
#shape of merged_inner1
merged_inner1.shape

(159, 9)

### **Question :** How usage of mobile services is affected by the manufacturer of user's mobile?

To answer this question we need to merge these datastes; first by merging "usage_df" and "user_df". 


**Merging usage_df with user_df**

In [23]:
##But we do not need all columns from the later one, so we can merge columns of two datasets selectively.
merged_left2 = pd.merge(usage_df, user_df[['use_id', 'platform', 'device']], on='use_id', how='left')

In [24]:
##overview on merged_left2
merged_left2.head()

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                    21.97                    4.82     1557.33   22787   
1                  1710.08                  136.88     7267.55   22788   
2                  1710.08                  136.88     7267.55   22789   
3                    94.46                   35.17      519.12   22790   
4                    71.59                   79.26     1557.33   22792   

  platform    device  
0  android  GT-I9505  
1  android  SM-G930F  
2  android  SM-G930F  
3  android     D2303  
4  android  SM-G361F

**Now we can match "merged_left2" with "device_df" using the "device" column which is common between them.**

In [25]:
## Renaming a column in the "Devices_df" dataframe to a more appropriate name
devices_df.rename(columns={"Retail Branding": "manufacturer"}, inplace=True)

  We need to add manufacturer columns from "devices_df" dataset to the "merged_left2" data frame.

In [26]:
merged_left3 = pd.merge(merged_left2,  devices_df[['manufacturer', 'Model']], left_on='device', right_on='Model',
how='left')

In [27]:
#overview on the new merged data frame
merged_left3.iloc[::30]

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id  \
0                      21.97                    4.82     1557.33   22787   
30                    324.27                   91.50      519.12   22830   
60                    249.26                  253.22     1557.33   22875   
90                    392.95                   89.48     1557.33   22916   
120                   250.87                  149.37      519.12   22945   
150                   215.92                   12.93     1557.33   22974   
180                   101.59                   84.41     5191.12   23018   
210                    42.75                   46.83     5191.12   23053   
240                   113.49                   48.16     6015.75   23838   
270                   677.02                  101.86     1132.23   24439   

    platform                  device manufacturer                   Model  
0    android                GT-I9505      Samsung                GT-I9505  
30   android                SM-G900F      Samsung                SM-G900F  
60   android                GT-I9505      Samsung                GT-I9505  
90   android                GT-I9195      Samsung                GT-I9195  
120  android          HTC Desire 626          HTC          HTC Desire 626  
150  android           Lenovo K51c78       Lenovo           Lenovo K51c78  
180  android              Moto G (4)     Motorola              Moto G (4)  
210  android  Vodafone Smart ultra 6          ZTE  Vodafone Smart ultra 6  
240      NaN                     NaN          NaN                     NaN  
270      NaN                     NaN          NaN                     NaN

just for demonstration, lets show all 'Samsung' devices in the 'devices_df'

In [28]:
devices_df[devices_df.manufacturer == 'Samsung']

manufacturer Marketing Name        Device      Model
9125       Samsung            NaN     GT-I5510M  GT-I5510M
9126       Samsung            NaN     GT-I5510T  GT-I5510T
9127       Samsung            NaN     GT-I5800L  GT-I5800L
9128       Samsung            NaN     GT-N7000B  GT-N7000B
9129       Samsung            NaN     GT-P7300B  GT-P7300B
...            ...            ...           ...        ...
10882      Samsung          VinsQ      SPH-M910   SPH-M910
10883      Samsung    VinsQ(M910)      SPH-M910   SPH-M910
10884      Samsung          W2016   royceltectc   SM-W2016
10885      Samsung          W2017  veyronltectc   SM-W2017
10886      Samsung          olleh           ik1  SMT-E5015

[1762 rows x 4 columns]

As a reult of merging process, now we are able to group users by their manufacturer of their device and find statistics which we were not able to find before merging.

In [29]:
#used ".agg" function to apply some aggregation across rows. 

merged_left4 = merged_left3.groupby("manufacturer").agg({ "outgoing_mins_per_month": "mean",
"outgoing_sms_per_month": "mean", "monthly_mb": "mean", "use_id": "count"})

Here we can see mobile usage, grouped by manufacturers.

In [30]:
merged_left4.rename(columns={"outgoing_mins_per_month": "mean_outgoing_mins_per_month",'outgoing_sms_per_month':'mean_outgoing_sms_per_month', 'monthly_mb': 'mean_monthly_mb', 'use_id' : 'count_use_id'}, inplace=True)

In [31]:
merged_left4.head()

mean_outgoing_mins_per_month  mean_outgoing_sms_per_month  \
manufacturer                                                              
HTC                             299.842955                    93.059318   
Huawei                           81.526667                     9.500000   
LGE                             111.530000                    12.760000   
Lava                             60.650000                   261.900000   
Lenovo                          215.920000                    12.930000   

              mean_monthly_mb  count_use_id  
manufacturer                                 
HTC               5144.077955            44  
Huawei            1561.226667             3  
LGE               1557.330000             2  
Lava             12458.670000             2  
Lenovo            1557.330000             2

### Adding two new columns representing total numbers for calls and sms for each manufacturer

In [32]:
#total numbers for calls
merged_left4['total_outgoing_mins_per_month'] = merged_left4['mean_outgoing_mins_per_month']*merged_left4['count_use_id']

In [33]:
#enting total numbers for sms
merged_left4['total_outgoing_sms_per_month'] = merged_left4['mean_outgoing_sms_per_month']*merged_left4['count_use_id']

In [34]:
#new data frame
merged_left4

mean_outgoing_mins_per_month  mean_outgoing_sms_per_month  \
manufacturer                                                              
HTC                             299.842955                    93.059318   
Huawei                           81.526667                     9.500000   
LGE                             111.530000                    12.760000   
Lava                             60.650000                   261.900000   
Lenovo                          215.920000                    12.930000   
Motorola                         95.127500                    65.666250   
OnePlus                         354.855000                    48.330000   
Samsung                         191.010093                    92.390463   
Sony                            177.315625                    40.176250   
Vodafone                         42.750000                    46.830000   
ZTE                              42.750000                    46.830000   

              mean_monthly_mb  count_use_id  total_outgoing_mins_per_month  \
manufacturer                                                                 
HTC               5144.077955            44                       13193.09   
Huawei            1561.226667             3                         244.58   
LGE               1557.330000             2                         223.06   
Lava             12458.670000             2                         121.30   
Lenovo            1557.330000             2                         431.84   
Motorola          3946.500000            16                        1522.04   
OnePlus           6575.410000             6                        2129.13   
Samsung           4017.318889           108                       20629.09   
Sony              3212.000625            16                        2837.05   
Vodafone          5191.120000             1                          42.75   
ZTE               5191.120000             1                          42.75   

              total_outgoing_sms_per_month  
manufacturer                                
HTC                                4094.61  
Huawei                               28.50  
LGE                                  25.52  
Lava                                523.80  
Lenovo                               25.86  
Motorola                           1050.66  
OnePlus                             289.98  
Samsung                            9978.17  
Sony                                642.82  
Vodafone                             46.83  
ZTE                                  46.83

In [35]:
#index of row with the maximum users
merged_left4['count_use_id'].argmax()

7

Statistics for the manufacturer with the highest numbner of users

In [36]:
merged_left4.loc[merged_left4['count_use_id'].idxmax()]

mean_outgoing_mins_per_month       191.010093
mean_outgoing_sms_per_month         92.390463
mean_monthly_mb                   4017.318889
count_use_id                       108.000000
total_outgoing_mins_per_month    20629.090000
total_outgoing_sms_per_month      9978.170000
Name: Samsung, dtype: float64

Statistics for the manufacturer with the highest "total_outgoing_mins_per_month"

In [37]:
merged_left4.loc[merged_left4['total_outgoing_mins_per_month'].idxmax()]


mean_outgoing_mins_per_month       191.010093
mean_outgoing_sms_per_month         92.390463
mean_monthly_mb                   4017.318889
count_use_id                       108.000000
total_outgoing_mins_per_month    20629.090000
total_outgoing_sms_per_month      9978.170000
Name: Samsung, dtype: float64

Statistics for the manufacturer with the highest "total_outgoing_sms_per_month"

In [38]:
merged_left4.loc[merged_left4['total_outgoing_sms_per_month'].idxmax()]

mean_outgoing_mins_per_month       191.010093
mean_outgoing_sms_per_month         92.390463
mean_monthly_mb                   4017.318889
count_use_id                       108.000000
total_outgoing_mins_per_month    20629.090000
total_outgoing_sms_per_month      9978.170000
Name: Samsung, dtype: float64

Let's add a columns which shows ranking of manufacturers with respect to "total_outgoing_mins_per_month"

In [39]:
merged_left4['call_Rank'] = merged_left4['total_outgoing_mins_per_month'].rank(ascending=False)

In [40]:
merged_left4.head(11)

mean_outgoing_mins_per_month  mean_outgoing_sms_per_month  \
manufacturer                                                              
HTC                             299.842955                    93.059318   
Huawei                           81.526667                     9.500000   
LGE                             111.530000                    12.760000   
Lava                             60.650000                   261.900000   
Lenovo                          215.920000                    12.930000   
Motorola                         95.127500                    65.666250   
OnePlus                         354.855000                    48.330000   
Samsung                         191.010093                    92.390463   
Sony                            177.315625                    40.176250   
Vodafone                         42.750000                    46.830000   
ZTE                              42.750000                    46.830000   

              mean_monthly_mb  count_use_id  total_outgoing_mins_per_month  \
manufacturer                                                                 
HTC               5144.077955            44                       13193.09   
Huawei            1561.226667             3                         244.58   
LGE               1557.330000             2                         223.06   
Lava             12458.670000             2                         121.30   
Lenovo            1557.330000             2                         431.84   
Motorola          3946.500000            16                        1522.04   
OnePlus           6575.410000             6                        2129.13   
Samsung           4017.318889           108                       20629.09   
Sony              3212.000625            16                        2837.05   
Vodafone          5191.120000             1                          42.75   
ZTE               5191.120000             1                          42.75   

              total_outgoing_sms_per_month  call_Rank  
manufacturer                                           
HTC                                4094.61        2.0  
Huawei                               28.50        7.0  
LGE                                  25.52        8.0  
Lava                                523.80        9.0  
Lenovo                               25.86        6.0  
Motorola                           1050.66        5.0  
OnePlus                             289.98        4.0  
Samsung                            9978.17        1.0  
Sony                                642.82        3.0  
Vodafone                             46.83       10.5  
ZTE                                  46.83       10.5

Now lets sort rows according to their "call_Rank"

In [41]:
merged_left4.sort_values('call_Rank', inplace = True)

In [42]:
merged_left4.head(11)

mean_outgoing_mins_per_month  mean_outgoing_sms_per_month  \
manufacturer                                                              
Samsung                         191.010093                    92.390463   
HTC                             299.842955                    93.059318   
Sony                            177.315625                    40.176250   
OnePlus                         354.855000                    48.330000   
Motorola                         95.127500                    65.666250   
Lenovo                          215.920000                    12.930000   
Huawei                           81.526667                     9.500000   
LGE                             111.530000                    12.760000   
Lava                             60.650000                   261.900000   
Vodafone                         42.750000                    46.830000   
ZTE                              42.750000                    46.830000   

              mean_monthly_mb  count_use_id  total_outgoing_mins_per_month  \
manufacturer                                                                 
Samsung           4017.318889           108                       20629.09   
HTC               5144.077955            44                       13193.09   
Sony              3212.000625            16                        2837.05   
OnePlus           6575.410000             6                        2129.13   
Motorola          3946.500000            16                        1522.04   
Lenovo            1557.330000             2                         431.84   
Huawei            1561.226667             3                         244.58   
LGE               1557.330000             2                         223.06   
Lava             12458.670000             2                         121.30   
Vodafone          5191.120000             1                          42.75   
ZTE               5191.120000             1                          42.75   

              total_outgoing_sms_per_month  call_Rank  
manufacturer                                           
Samsung                            9978.17        1.0  
HTC                                4094.61        2.0  
Sony                                642.82        3.0  
OnePlus                             289.98        4.0  
Motorola                           1050.66        5.0  
Lenovo                               25.86        6.0  
Huawei                               28.50        7.0  
LGE                                  25.52        8.0  
Lava                                523.80        9.0  
Vodafone                             46.83       10.5  
ZTE                                  46.83       10.5

As we can see two rows have 10.5 ranke. Let's fix that

In [43]:
merged_left4["call_Rank"] = merged_left4['call_Rank'].round().astype(int)

In [44]:
merged_left4.head(11)

mean_outgoing_mins_per_month  mean_outgoing_sms_per_month  \
manufacturer                                                              
Samsung                         191.010093                    92.390463   
HTC                             299.842955                    93.059318   
Sony                            177.315625                    40.176250   
OnePlus                         354.855000                    48.330000   
Motorola                         95.127500                    65.666250   
Lenovo                          215.920000                    12.930000   
Huawei                           81.526667                     9.500000   
LGE                             111.530000                    12.760000   
Lava                             60.650000                   261.900000   
Vodafone                         42.750000                    46.830000   
ZTE                              42.750000                    46.830000   

              mean_monthly_mb  count_use_id  total_outgoing_mins_per_month  \
manufacturer                                                                 
Samsung           4017.318889           108                       20629.09   
HTC               5144.077955            44                       13193.09   
Sony              3212.000625            16                        2837.05   
OnePlus           6575.410000             6                        2129.13   
Motorola          3946.500000            16                        1522.04   
Lenovo            1557.330000             2                         431.84   
Huawei            1561.226667             3                         244.58   
LGE               1557.330000             2                         223.06   
Lava             12458.670000             2                         121.30   
Vodafone          5191.120000             1                          42.75   
ZTE               5191.120000             1                          42.75   

              total_outgoing_sms_per_month  call_Rank  
manufacturer                                           
Samsung                            9978.17          1  
HTC                                4094.61          2  
Sony                                642.82          3  
OnePlus                             289.98          4  
Motorola                           1050.66          5  
Lenovo                               25.86          6  
Huawei                               28.50          7  
LGE                                  25.52          8  
Lava                                523.80          9  
Vodafone                             46.83         10  
ZTE                                  46.83         10